### ISSUE TITLE
Using ransackable_attributes with Active Admin errors

### ISSUE THREAD

**Issue by viktorsmari** [activeadmin/activeadmin#7035]
We have a Rails 6.1 app with `activeadmin` 2.9.0 and `ransack` 2.4.2 and when we visit `/admin/users` we get the following error:

**ArgumentError in Admin::Users#index**

**Polymorphic associations do not support computing the class.**

The error goes away if we remove the following section:

```ruby
#app/models/user.rb
def self.ransackable_attributes(auth_object = nil)
  super & %w(first_name last_name slug)
end
```

We need this section to prevent users being able to guess (ransack) someone else's password or email.

Is there a better way to whitelist which User columns are searchable or to prevent this error?

The repo is open source if you want to try to reproduce:
https://github.com/fablabbcn/makeworks

**Comment by thdurante** [comment]
hey, @viktorsmari did you find a solution for your problem?

**Comment by viktorsmari** [comment]
No, sorry @thdurante - would be great though to get a solution for this!

**Comment by thdurante** [comment]
I think I got it working @viktorsmari.

Basically I had two different scenarios:

- Ransack search via activeadmin - I did not want any limitations here - ie.: search against all attrs
- Ransack search via API - I wanted to whitelist the fields the user was able to search against

The problem was that I had my method defined as follows for one of my models:

```
class ModelX < ApplicationRecord
  [...]

  private

  def self.ransackable_attributes(_auth_object = nil)
    SEARCH_FIELDS.map(&:to_s) + _ransackers.keys
  end
end
```

That worked for the API but not for ActiveAdmin so I was getting exactly the same error as you because the attributes were not whitelisted when I was filtering through the admin. I changed the `ransackable_attributes` to:

```
def self.ransackable_attributes(auth_object = nil)
  return SEARCH_FIELDS.map(&:to_s) + _ransackers.keys if auth_object.try(:==, :api)
  
  super
end
```

And then changed my API to set an auth object like the following:

```
model_relation.ransack({ name_i_cont: query[:search_term] }, { auth_object: :api }).result
```

So now I account for different scenarios/situations when whitelisting the attributes on my model and that solved the issue.
Let me know if that's helpful.
